src/testing/collector.py

```diff
@@ -0,0 +1,70 @@
+import os
+from dataclasses import dataclass
+from pathlib import Path
+from typing import List, Optional, Pattern
+
+from starkware.cairo.lang.compiler.preprocessor.preprocessor_error import (
+    PreprocessorError,
+)
+
+from src.starknet_compilation import StarknetCompiler
+from src.testing.reporter import TestReporter
+from src.testing.utils import collect_subdirectories, TestSubject
+
+
+class CollectionError(Exception):
+    pass
+
+
+@dataclass
+class TestCollector:
+    sources_directory: Path
+    include_paths: Optional[List[str]] = None
+
+    # TODO: Optimize, by returning preprocessed test program, for reuse when compiling it for test runs
```

Comment:
Each TODO should have a corresponding task.

Reply:
Good point :)